language/move-vm/runtime/src/interpreter.rs

```diff
@@ -246,6 +273,55 @@ impl Interpreter {
         }
     }
 
+    /// Make sure only private/friend function can only be invoked by modules under the same address.
+    fn check_friend_or_private_call(
+        &self,
+        caller: &Arc<Function>,
+        callee: &Arc<Function>,
+    ) -> VMResult<()> {
+        if callee.is_friend_or_private() {
+            match (caller.module_id(), callee.module_id()) {
+                (Some(caller_id), Some(callee_id)) => {
+                    if caller_id.address() == callee_id.address() {
```

Comment:
This isn't inherent to friends. Like sure we have this check in friends right now, but it might not always be the case. I would probably check if the module is in the friend list?

It also doesn't catch invalid private access from other addresses